Implement get_fattime

Hrushi20 opened the pull request:
Pr solves Issue #159 by adding logic to write and read time from DS1307 module.

This is done by reading 32-bit unix timestamp from UART1TX register, converting it to bcd format and storing it on RTC chip via I2C protocol.

Hrushi20 commented:
Hey, I am using a Mac M1 chip. Is there support for xc-16 compilter. I am not able to compile and check if the build succeeds or not.

bessman commented:
> Hey, I am using a Mac M1 chip. Is there support for xc-16 compilter. I am not able to compile and check if the build succeeds or not. I tried downloading the xc-16 compiler. The cmake build fails: `No xc-16 compiler found`

I am unsure how to get this working as I don't use Mac. Here is a Microchip forum thread where some people seem to have at least gotten xc8 working on M1: https://forum.microchip.com/s/topic/a5C3l000000Me63EAC/t384186

CMake uses a [cmake-microchip](https://github.com/fossasia/cmake-microchip) toolchain file to detect xc16. On Mac, it searches in `/Applications/microchip/`. Make sure that's where you installed xc16.

Hrushi20 commented:
I didn't not add parameters to `RTC_SetTime`, I am reading the unix_timestamp using UART1 protocol. 

I did not understand the Wrapper functions which are to be added in commands.c file. Can you give me a brief overview of what the commands.c file does?

bessman commented:
> I didn't not add parameters to `RTC_SetTime`, I am reading the unix_timestamp using UART1 protocol sent by the host.
> 
> I did not understand the Wrapper functions which are to be added in commands.c file. Can you give me a brief overview of what the commands.c file does?

The main point of interest in `commands.c` is [`cmd_table`](https://github.com/fossasia/pslab-firmware/blob/e0e82dfc963ff9860eafe636dfcf9197ad7e4824/src/commands.c#L73). It is a 2D jump table, containing function pointers which are run in response to command instructions from the host.

When the host sends a command to the PSLab, the first two bytes are an index into this table (we refer to these bytes as "primary command" and "secondary command"). For example, if the host sends `[13, 1]`, it means "run the function at `cmd_table[13][1]`".

The functions pointed to by `cmd_table` must have type `response_t (*)(void)`. They take no parameters, because the functions themselves are responsible for collecting any additional data they may require from the host using `UART_Read*`. However, this means these functions cannot be used internally in the firmware. For example, we want to be able to read the current time from the FatFS driver, but we cannot use `RTC_GetTime` because it will try to send the result to a host via UART.

Therefore, if we have a function which we want to use both internally within pslab-firmware and over serial from a host, we create a wrapper:

``` c
response_t do_something(uint8_t a, uint16_t* bp) {
    // Do something with a and put the result in b.
}

response_t do_something_serial(void) {
    uint8_t a = UART1_Read(); // Read input from host.
    uint16_t b = 0;
    response_t res = do_something(a, &b);
    UART1_WriteInt(b); // Write output to host.
    return res;
}
```

Hrushi20 commented:
Hey, Thanks for the explanation 😄. I've added the wrapper function and updated the commands.c file. What do I return to host after setting time to rtc?

Also, while importing the `rtc.h` header into `ff_time.c`, the build is failing. 
I imported the file using `#include "../../helpers/rtc.h"`

Thanks

bessman commented:
> What do I return to host after setting time to rtc?

Nothing. The return value (i.e. what is actually `return`ed, not sent to the host) from a command function is ultimately sent to the host by the main state machine, [here](https://github.com/fossasia/pslab-firmware/blob/e0e82dfc963ff9860eafe636dfcf9197ad7e4824/src/states.c#L34).

> Also, while importing the `rtc.h` header into `ff_time.c`, the build is failing. I imported the file using `#include "../../helpers/rtc.h"`

Yes, that is how we handle includes.

bessman commented on src/helpers/rtc.c:
```diff
@@ -8,24 +9,72 @@
+    uint32_t unix_timestamp = UART1_read_u32();
+
+    response_t res = RTC_SetTime(&unix_timestamp);
+    UART1_Write(res);
```
The command function should not sent its acknowledge byte to the host directly; this is handled in the main loop.

Hrushi20 commented:
> > What do I return to host after setting time to rtc?
> 
> Nothing. The return value (i.e. what is actually `return`ed, not sent to the host) from a command function is ultimately sent to the host by the main state machine, [here](https://github.com/fossasia/pslab-firmware/blob/e0e82dfc963ff9860eafe636dfcf9197ad7e4824/src/states.c#L34).
> 
> > Also, while importing the `rtc.h` header into `ff_time.c`, the build is failing. I imported the file using `#include "../../helpers/rtc.h"`
> 
> Yes, that is how we handle includes.

Why is the build failing? How do I fix the issue of importing rtc.h header file?

Hrushi20 commented:
<img width="1349" alt="Screenshot 2023-11-17 at 3 23 55 AM" src="https://github.com/fossasia/pslab-firmware/assets/65367566/07572a41-c900-49fe-9ce5-519c3fb38ca4">

The build fails when I import the `rtc.h` file into `ff_time.c`

bessman commented:
Ah, sorry, I misunderstood. It fails because `rtc.h` is missing an include: `#include "../commands.h"`. Not your fault; it should have already been included I guess everything that included `rtc.h` before now already included `commands.h` themselves, so we didn't notice it was missing.

Hrushi20 commented:
Thanks. I've updated the example using rtc_gettime function.

bessman commented on src/commands.c:
```diff
@@ -279,7 +279,7 @@ command_func_t* const cmd_table[NUM_PRIMARY_CMDS + 1][NUM_SECONDARY_CMDS_MAX + 1
     { // 12 SENSORS
      // 0                               1 RTC_SETTIME                   2 RTC_SETDIGIT                  3 RTC_GETTIME
-        Undefined,                      RTC_SetTime,                    RTC_SetDigit,                   RTC_GetTime,
+        Undefined,                      RTC_CmdSetTime,                    RTC_SetDigit,                   RTC_CmdGetTime,
```
```suggestion
        Undefined,                      RTC_CmdSetTime,                 RTC_SetDigit,                   RTC_CmdGetTime,
```

bessman replied:
Adjust whitespace to align table.

bessman commented on src/helpers/rtc.c:
```diff
@@ -8,24 +9,71 @@
-    buffer[6] = UART1_Read();         // month
-    buffer[7] = UART1_Read();         // year
-    buffer[8] = UART1_Read();         // control
+    buffer[1] = data_to_bcd(sec) & ~(1<<7);                      // seconds
```
I'm not sure about `~(1<<7)`. It evaluates to -129, which has the same bit representation as 0x7F, but bitwise operations with signed values should be avoided.

bessman replied:
What you're doing is correct, but the way you're doing it is less readable than it could be.

Instead of using `~(1<<7)`, create a variable `uint8_t oscillator_enable = 0x7F`, with a comment explaining its purpose.

bessman commented on src/helpers/rtc.c:
```diff
@@ -8,24 +9,71 @@
+    return left & right;
+}
+
+response_t RTC_SetTime(uint32_t *unix_timestamp) {
```
```suggestion
response_t RTC_SetTime(uint32_t const* const unix_timestamp) {
```

bessman commented on src/helpers/rtc.h:
Please add/update function documentation where necessary.

bessman commented on src/sdcard/fatfs/ff_time.c:
Please remove the commented-out code.

bessman reviewed (changes requested):
Hi, I haven't had time to try this out yet, but here are some general comments.

bessman commented on src/helpers/rtc.c:
```diff
@@ -8,24 +9,73 @@
-response_t RTC_SetTime(void) {
-    
+// This bit is used to enable the oscillator. When negated, the oscillator is disabled.
+uint8_t oscillator_enable = 0x7F
```
```suggestion
static uint8_t oscillator_enable = 0x7F
```

bessman commented on src/helpers/rtc.c:
```diff
@@ -8,24 +9,73 @@
+// This bit is used to enable the oscillator. When negated, the oscillator is disabled.
+uint8_t oscillator_enable = 0x7F
+
+uint8_t data_to_bcd(uint8_t data){
```
```suggestion
static uint8_t data_to_bcd(uint8_t data){
```

bessman commented on src/helpers/rtc.c:
```diff
@@ -8,24 +9,73 @@
+    return bcd;
+}
+
+uint8_t bcd_to_data(uint8_t bcd){
```
```suggestion
static uint8_t bcd_to_data(uint8_t bcd){
```

bessman commented on src/sdcard/fatfs/ff_time.c:
```diff
@@ -59,23 +59,27 @@
-DWORD get_fattime (void)
-{
-    return decimalToFatTime(2018, 6, 31, 5, 10, 30);
 }
```
```suggestion
}

```

bessman reviewed (changes requested):
Almost there :)

bessman commented:
I will test this tonight.

Did you get the message I sent you on Matrix? I set up a remote dev environment where you can test your changes on a PSLab.

bessman commented:
I refined the script a bit:

```python
import enum
import logging
import time

import pytest

import pslab


class Acknowledge(enum.Enum):
    SUCCESS = 1
    ARGUMENT_ERROR = 2
    FAILURE = 3


logger = logging.getLogger(__name__)


def set_time(psl: pslab.ScienceLab, ts: int) -> Acknowledge:
    psl.send_byte(13)  # SENSORS
    psl.send_byte(1)  # RTC_SETTIME
    logger.debug("Sent command: RTC_SETTIME")
    ser_ts = pslab.protocol.Integer.pack(ts)  # Serialize timestamp
    logger.debug(f"Serialized timestamp: {ts} -> {list(ser_ts)}")
    psl.write(ser_ts)  # Send serialized timestamp to PSLab
    logger.debug("Wrote timestamp to PSLab")
    ack = Acknowledge(psl.get_byte())
    logger.debug(f"Got ACK byte: {ack.name}")
    return ack


def get_time(psl: pslab.ScienceLab) -> int:
    psl.send_byte(13)  # SENSORS
    psl.send_byte(3)  # RTC_GETTIME
    logger.debug("Sent command: RTC_GETTIME")
    ser_ts = psl.read(4)  # Read serialized timestamp from PSLab
    logger.debug(f"Got serialized timestamp: {list(ser_ts)}")
    ack = Acknowledge(psl.get_ack())
    logger.debug(f"Got ACK byte: {ack.name}")
    ts = pslab.protocol.Integer.unpack(ser_ts)[0]  # Deserialize timestamp
    logger.debug(f"Deserialzed timestamp: {list(ser_ts)} -> {ts}")
    return ts


def test_rtc():
    logging.basicConfig(level=logging.DEBUG)
    psl = pslab.ScienceLab()
    logger.debug("PSLab connected")
    ts_in = int(time.time())
    logger.debug(f"Current Unix time: {ts_in}")
    # Set current time
    logger.debug("Setting time on RTC")
    ack = set_time(psl, ts_in)
    assert ack == Acknowledge.SUCCESS
    # Wait one second
    logger.debug("Waiting for one second")
    time.sleep(1)
    # Readback time
    logger.debug("Reading back time from PSLab")
    ts_out = get_time(psl)
    assert ts_out == ts_in + 1
```

I've added this script to the remote dev environment. To run it, just run `pytest` after logging in. When the test passes, the RTC setter and getter functions are OK.

bessman commented on src/helpers/rtc.c:
```diff
@@ -8,24 +9,73 @@
+    uint8_t date = tm_info->tm_mday;
+    uint8_t month = tm_info->tm_mon + 1;
+
+    if(tm_info->tm_year < 2000 || tm_info->tm_year > 2099)
```
I think this test is wrong. `tm_year` is years since 1900 (which you account for just below), so this test will be true if the year is less than 3900 or greater than 3999.

Hrushi20 commented:
I got the tes working. Just testing it on the raspberry pi.

Hrushi20 commented:
Can you review the ff_time code? I changed the logic there based on unix timestamp.

bessman commented:
> I got the test working. Just tested it on the raspberry pi.

The RTC test isn't passing for me:

```
________________________________________________ test_rtc ________________________________________________

    def test_rtc():
        psl = pslab.ScienceLab()
        logger.debug("PSLab connected")
        ts_in = int(time.time())
        logger.debug(f"Current Unix time: {ts_in}")
        # Set current time
        logger.debug("Setting time on RTC")
        ack = set_time(psl, ts_in)
        assert ack == Acknowledge.SUCCESS
        # Wait one second
        logger.debug("Waiting for one second")
        time.sleep(1)
        # Readback time
        logger.debug("Reading back time from PSLab")
        ts_out = get_time(psl)
>       assert ts_out == ts_in + 1
E       assert 1701649124 == (1701721123 + 1)

test_rtc.py:60: AssertionError
------------------------------------------- Captured log call --------------------------------------------
INFO     pslab.serial_handler:serial_handler.py:215 Connected to PSLab V6
 on /dev/ttyUSB0.
DEBUG    test_rtc:test_rtc.py:47 PSLab connected
DEBUG    test_rtc:test_rtc.py:49 Current Unix time: 1701721123
DEBUG    test_rtc:test_rtc.py:51 Setting time on RTC
DEBUG    test_rtc:test_rtc.py:22 Sent command: RTC_SETTIME
DEBUG    test_rtc:test_rtc.py:24 Serialized timestamp: 1701721123 -> [35, 52, 110, 101]
DEBUG    test_rtc:test_rtc.py:26 Wrote timestamp to PSLab
DEBUG    test_rtc:test_rtc.py:28 Got ACK byte: SUCCESS
DEBUG    test_rtc:test_rtc.py:55 Waiting for one second
DEBUG    test_rtc:test_rtc.py:58 Reading back time from PSLab
DEBUG    test_rtc:test_rtc.py:35 Sent command: RTC_GETTIME
DEBUG    test_rtc:test_rtc.py:37 Got serialized timestamp: [228, 26, 109, 101]
DEBUG    test_rtc:test_rtc.py:39 Got ACK byte: SUCCESS
DEBUG    test_rtc:test_rtc.py:41 Deserialzed timestamp: [228, 26, 109, 101] -> 1701649124
======================================== short test summary info =========================================
FAILED test_rtc.py::test_rtc - assert 1701649124 == (1701721123 + 1)
=========================================== 1 failed in 2.81s ============================================
```

bessman commented:
> Can you review the ff_time code? I changed the logic there based on unix timestamp.

I've added another test script, `test_fattime.py`, which you can run in the same way as the other test script. It sets the RTC time, touches a file on the SD-card, and then reads the file's modification timestamp.

It relies on some recently added SD-card functionality, so you will need to rebase your branch on main to get it to work.

You will also need to run cmake with the `-DLEGACY_HARDWARE=1` flag. This is because the PSLab which is connected to the RPi is a prototype of the upcoming v6 hardware. (This wasn't necessary before because the RTC is the same in both hardware versions. The SD-card, on the other hand, is implemented differently between hardware versions)

bessman commented on src/sdcard/fatfs/ff_time.c:
```diff
@@ -59,23 +59,27 @@
+    fatTime |= ( ((DWORD)tm_info->tm_min) << 5 );
+    fatTime |= ( ((DWORD)tm_info->tm_hour) << 11 );
+    fatTime |= ( ((DWORD)tm_info->tm_wday + 1) << 16 );
+    fatTime |= ( ((DWORD)tm_info->tm_mday) << 21 );
```
This should be the month. Use `tm_info->tm_mon + 1`.

bessman commented on src/sdcard/fatfs/ff_time.c:
```diff
@@ -59,23 +59,27 @@
+    fatTime |= ( ((DWORD)tm_info->tm_hour) << 11 );
+    fatTime |= ( ((DWORD)tm_info->tm_wday + 1) << 16 );
+    fatTime |= ( ((DWORD)tm_info->tm_mday) << 21 );
+    fatTime |= ( ((DWORD)(tm_info->tm_year + 1900)) << 25 );
```
`tm_info->tm_year` is years since 1900, but FatFS wants years since 1980. This should be `tm_info->tm_year - 80`

bessman commented on src/helpers/rtc.c:
```diff
@@ -8,24 +9,69 @@
-    buffer[8] = UART1_Read();         // control
+    buffer[1] = data_to_bcd(sec) & oscillator_enable;            // seconds
+    buffer[2] = data_to_bcd(min);                                // minutes
+    buffer[3] = (data_to_bcd(hours) | 0x20);                     // hours (hrs format)
```
`0x20` is the fifth bit (zero indexed). You should not set this bit, as it is used to store the second 10h increment in 24h mode.

There is no need for any special handling of hours in 24h mode.

bessman commented on src/helpers/rtc.c:
```diff
@@ -37,20 +83,42 @@ response_t RTC_SetDigit(void) {
+        // Need to convert from bcd to int.
+        tm_info.tm_sec = bcd_to_data(buffer[0]);
+        tm_info.tm_min = bcd_to_data(buffer[1]);
+        tm_info.tm_hour = bcd_to_data(buffer[2] & 0x1F); // Remove the Hour Factor.
```
```suggestion
        tm_info.tm_hour = bcd_to_data(buffer[2]);
```

Hrushi20 commented:
> > I got the test working. Just tested it on the raspberry pi.
> 
> The RTC test isn't passing for me:
> 
> ```
> ________________________________________________ test_rtc ________________________________________________
> 
>     def test_rtc():
>         psl = pslab.ScienceLab()
>         logger.debug("PSLab connected")
>         ts_in = int(time.time())
>         logger.debug(f"Current Unix time: {ts_in}")
>         # Set current time
>         logger.debug("Setting time on RTC")
>         ack = set_time(psl, ts_in)
>         assert ack == Acknowledge.SUCCESS
>         # Wait one second
>         logger.debug("Waiting for one second")
>         time.sleep(1)
>         # Readback time
>         logger.debug("Reading back time from PSLab")
>         ts_out = get_time(psl)
> >       assert ts_out == ts_in + 1
> E       assert 1701649124 == (1701721123 + 1)
> 
> test_rtc.py:60: AssertionError
> ------------------------------------------- Captured log call --------------------------------------------
> INFO     pslab.serial_handler:serial_handler.py:215 Connected to PSLab V6
>  on /dev/ttyUSB0.
> DEBUG    test_rtc:test_rtc.py:47 PSLab connected
> DEBUG    test_rtc:test_rtc.py:49 Current Unix time: 1701721123
> DEBUG    test_rtc:test_rtc.py:51 Setting time on RTC
> DEBUG    test_rtc:test_rtc.py:22 Sent command: RTC_SETTIME
> DEBUG    test_rtc:test_rtc.py:24 Serialized timestamp: 1701721123 -> [35, 52, 110, 101]
> DEBUG    test_rtc:test_rtc.py:26 Wrote timestamp to PSLab
> DEBUG    test_rtc:test_rtc.py:28 Got ACK byte: SUCCESS
> DEBUG    test_rtc:test_rtc.py:55 Waiting for one second
> DEBUG    test_rtc:test_rtc.py:58 Reading back time from PSLab
> DEBUG    test_rtc:test_rtc.py:35 Sent command: RTC_GETTIME
> DEBUG    test_rtc:test_rtc.py:37 Got serialized timestamp: [228, 26, 109, 101]
> DEBUG    test_rtc:test_rtc.py:39 Got ACK byte: SUCCESS
> DEBUG    test_rtc:test_rtc.py:41 Deserialzed timestamp: [228, 26, 109, 101] -> 1701649124
> ======================================== short test summary info =========================================
> FAILED test_rtc.py::test_rtc - assert 1701649124 == (1701721123 + 1)
> =========================================== 1 failed in 2.81s ============================================
> ```

Hey, can you flash and try again. Are you using the latest commit? The test is passing for me.

Hrushi20 commented:
<img width="684" alt="Screenshot 2023-12-05 at 9 47 23 AM" src="https://github.com/fossasia/pslab-firmware/assets/65367566/fc8cc0a7-96c6-4b71-8a51-67ee0a7d9cf1">

Hrushi20 commented:
The test_fattime fails. There is an issue with the hours format. I'll debug it during ny free time.

bessman commented:
> Hey, can you flash and try again. Are you using the latest commit? The test is passing for me.

Now it's working :+1: 

> The test_fattime fails. There is an issue with the hours format. I'll debug it during ny free time.

It was a problem with the test; It didn't take the timezone into account. I've fixed it and the both tests are passing!

I'll give this a final look tonight, but it seems ready for merge.

Hrushi20 commented:
I'm looking forward to the merge 😁
